### PR TITLE
ENH: stats: Add `nan_policy` optional argument for `stats.rankdata` 

### LINKED ISSUE #11790: NaN handling of stats.rankdata
I think NaN handling of stats.rankdata is not correct. 

#### Reproducing code example:

Current code is

```
rankdata([0, 2, 3, 2, np.nan, np.nan])
> [1.  2.5 4.  2.5 5.  6. ]
```

However, I think it should be

```
rankdata([0, 2, 3, 2, np.nan, np.nan])
> [1.  2.5 4.  2.5 np.nan  np.nan ]
```

or 

return nan

or

raise an exception.

#### Scipy/Numpy/Python version information:
Python: 3.7.7
scipy: 1.7.1

### PR DESCRIPTION
**Pull request by AtsushiSakai**
#### Reference issue

fix #11790

#### What does this implement/fix?

I added `nan_policy` optional argument for `stats.rankdata` based on [A Design Specification for nan\_policy](http://scipy.github.io/devdocs/dev/api-dev/nan_policy.html) and its test.

#### Additional information

This PR is a reviving #13215

### CODE REVIEW

File: scipy/stats/tests/test_rank.py
Comment (mdhaber) [inline scipy/stats/tests/test_rank.py:200]:
```suggestion
        data = [0, 2, 3, -2, np.nan, np.nan]
        res = rankdata(data, nan_policy="propagate")
        assert_array_equal(res, [2., 3., 4., 1., 5., 6.])
```
`data` can be re-used below.
The order of inputs to `assert_array_equal` doesn't really matter in practice, but this reverses them to adhere to the convention:
image

This is also nice because there is no more line continuation / looks cleaner IMO.

**Review by mdhaber — COMMENTED** [review]
Nice! This looks consistent with

image

It relies on the implementation always ranking NaNs last. Is that true for all the different comparison methods?

I'd like to write a more comprehensive property-based test before merging, but this looks close to me.

**Comment by AtsushiSakai** [comment]
Thank you for reviewing.

> Is that true for all the different comparison methods?

Do you mean the comparison methods are the `method` arguments?
image

> I'd like to write a more comprehensive property-based test

Do you mean the property-based test is using @pytest.mark.parametrize?

File: scipy/stats/tests/test_rank.py
Comment (mdhaber) [inline scipy/stats/tests/test_rank.py:195]:
```suggestion
    methods = ['average', 'min', 'max', 'dense', 'ordinal']

    @pytest.mark.parametrize('method', methods)
    def test_nan_policy(self, method):
```

I tried making this change (and using `method=method` in calls to `rankdata` below) and found that tests fail for all methods except 'average'. This is pretty clear from the logic. I'd suggest refactoring so that there is only one `return` statement (at the end).

**Review by mdhaber — COMMENTED** [review]
A few comments upon closer review:

- Please consider `method` other than `average` (and parametrize the test to show that it works for all methods).
- To fully address the issue, I don't think that `nan_policy='propagate'` can just run the code as-is. We should treat all NaNs as identical like we now do in `mode`. Therefore, all NaNs should have the same rank except for when `method='ordinal'`. See gh-15516 for related discussion.

Regarding the property-based test, here is an example. The point is that the test does not just check the output for particular values of the input; it uses arbitrary data and NaN patterns and checks that the required properties are satisfied: the behavior is as though NaNs had been remove moved from each axis-slice before ranks were assigned, and the nan-patttern of the output is the same as the nan-pattern of the input.

<details>

```python3
    @pytest.mark.parametrize('axis', range(2))
    @pytest.mark.parametrize('method', methods)
    def test_nan_policy_omit_3d(self, axis, method):

        shape = (20, 21, 22)
        rng = np.random.default_rng(abs(hash('falafel')))

        a = rng.random(size=shape)
        i = rng.random(size=shape) < 0.4
        j = rng.random(size=shape) < 0.1
        k = rng.random(size=shape) < 0.1
        a[i] = np.nan
        a[j] = -np.inf
        a[k] - np.inf

        def rank_1d_omit(a, method):
            out = np.zeros_like(a)
            i = np.isnan(a)
            a_compressed = a[~i]
            res = rankdata(a_compressed, method)
            out[~i] = res
            out[i] = np.nan
            return out

        def rank_omit(a, method, axis):
            return np.apply_along_axis(lambda a: rank_1d_omit(a, method), axis, a)

        res = rankdata(a, method, axis=axis, nan_policy='omit')
        res0 = rank_omit(a, method, axis=axis)

        assert_array_equal(res, res0)
```

</details>

**Comment by AtsushiSakai** [comment]
Thank you for the detailed explanations. I think `omit` and `raise` are supported now.

> We should treat all NaNs as identical like we now do in mode
I agree. But I'm not sure NaN's rank should be the biggest rank or smallest, or NaN in `propagate`.
I think it should be NaN in `propagate`. This means `omit` and `propagate` will be same behavior.
What do you think?

**Comment by mdhaber** [comment]
It would also be reasonable if `propagate` were to cause all other elements to be NaN, actually: if a NaN's rank relative to another element is undefined, then the other element's rank relative to the NaN is undefined.

**Comment by AtsushiSakai** [comment]
OK. I would like to hear other options. Current options are: 

input:
rankdata([0, 2, 3, 2, np.nan, np.nan], nan_policy='propagate')

output options:
* 1: [1.  2.5 4.  2.5 5.  5. ]
* 2: [2.  3.5 5.  3.5 1.  1. ]
* 3: [1.  2.5 4.  2.5 nan.  nan. ]
* 4: [nan  nan nan  nan nan.  nan. ]

**Comment by tirthasheshpatel** [comment]
First thoughts: I agree with @AtsushiSakai. Keeping `nan`'s rank as `nan` when `nan_policy='propagate'` is more intuitive (option 3). I also had similar thoughts for `stats.mode` where too the behavior of `nan_policy='propagate'` and `nan_policy='omit'` became exactly identical.

After thinking about it a bit, it also makes sense to just return all `nan`s since `nan`s "propagate". A middle ground between expected and correct behavior would be to change the default value of `nan_policy` to `'omit'` so the users get the intuitive result by default. We can then change the output of `nan_policy='propagate'` to return all `nan` (option 5) and add a note in docs so users opting for it know what they would get. What do you think @mdhaber @AtsushiSakai?

**Comment by AtsushiSakai** [comment]
Thank you for your comment. I like this idea:
> just return all nans since nans "propagate". change the default value of `nan_policy` to 'omit'

**Comment by AtsushiSakai** [comment]
CI failures looks `_contains_nan` starts using `isnan` in #16209, and `isnan` does not support string type array. I will consider a good solution.

**Comment by mdhaber** [comment]
@AtsushiSakai Would you post that fix of `_contains_nan` as a separate PR and add a unit test? It is conceptually distinct from this PR, and I want to make sure it makes it into the next release even if this doesn't. When that new PR merged into `main`, please merge `main` here (not rebase). Thanks!

**Review by mdhaber — APPROVED** [review]
@AtsushiSakai I edited the docstring a bit, but otherwise, this looks good.

If tests pass, I think this is ready to go. Ideally, the tests for `raise` and `propagate` would be as strong as the test for `omit`, but it would probably be overkill.

File: scipy/stats/tests/test_rank.py
Comment (tupui) [inline scipy/stats/tests/test_rank.py:199]: Brilliant 😂

Comment (mdhaber) [reply scipy/stats/tests/test_rank.py:199]: Hah thanks : ) 

Quick question on this one - I don't need you to review the code, but I just wanted another opinion on the behavior.

Before this PR, NaNs were assigned real ranks greater than all other ranks. 

Now:
image

Can we consider the original behavior a bug since it wasn't documented and isn't really meaningful, so we can change the default behavior a bit here? Currenlly 'omit' is the default because it minimizes impact, but could we change it so that 'propagate' is the default? Do you agree with what 'propagate' does?

Comment (tupui) [reply scipy/stats/tests/test_rank.py:199]: Wouldn't it be surprising to return nans instead of something for a user? TBH it's difficult for me to judge as I am almost never working with nans. But I think I would expect the function to deal with it and returns something instead of bailing out with nans.

Comment (mdhaber) [reply scipy/stats/tests/test_rank.py:199]: _Note: the image above was wrong before. It showed the documentation for `method`, but it was supposed to show the documentation for the new `nan_policy`._

That's OK. I thought you might have an opinion since we've worked together on other NaN PRs.
If it seems surprising to the user, think about how surprising NaN inputs are to the poor function! How is it to divine whether the user considers `np.nan` to be less than or greater than any given number? : )
So `rankdata` will deal with `x = [0, np.nan]` analogously to `np.min` when `nan_policy='propagate'` and analogously to `np.nanmin` when `nan_policy='omit'`.

I'll go ahead and merge since there are three in agreement and one undecided, but I'll open up a related PR so there will still be a chance for others to review. Thanks @AtsushiSakai @tirthasheshpatel @tupui!

**Pull request merged**
